Import album extras

tty418 opened the pull request:
#### Database Migration
YES - Constraint removed from ExtraFiles table to enable for album-wide extras to be inserted.

#### Motivation
Music collections often include additional extra files for each album such as metadata (.nfo's, EAC .log files, .txt files etc), album art images, as well as .cue sheets. Currently, these files get lost during import - both during manual import in-place, as well as from the download client. 

My goal was for this feature to work seamlessly, not break existing flows, and be **on by default**. The lack of album extras 'surprises' quite a few people, and often the revelation comes a bit late (e.g. after processing half my existing music collection),

#### Description
This PR adds handling for extra files which are specific to a whole album and not a single track. The extra files themselves will not be renamed and will be kept as-is during import. 

These 'album extras' will be handled **only if renaming is enabled**. Another requirement is that an **album directory is present** in the track naming template. If these two conditions are not satisfied, importing the album extras will not be a safe operation due to conflicts with extras from other albums.

#### Implementation details (relevant for the PR)
The current extra file handling is geared towards per-track extras. Adding support for album extras is impossible without having to touch a ton of working code and redefine interfaces. The new feature has a separate implementation as it follows completely different logic. It is easier to follow and won't break anything.

The list of relevant file extensions is configurable in the UI. The defaults were changed to something more suitable for a music service.

#### Implementation todo's (draft PR):
- [x] Test migration against PostgreS
- [x] Check file naming configuration for the presence of an album dir
- [x] Retest with downloads and manual imports after refactoring
- [x] Test rename (move)
- [x] Avoid comparing source/destination paths?

#### Todos
- [x] Tests
- [x] Translation Keys (./src/NzbDrone.Core/Localization/Core/en.json)
- [ ] [Wiki Updates](https://wiki.servarr.com)

#### Issues Fixed or Closed by this PR

* Fixes #484

tty418 committed:
Import album extras for manual imports and downloads

tty418 commented on src/NzbDrone.Core/Extras/Files/AlbumExtraFileManager.cs:
```diff
@@ -0,0 +1,164 @@
+            var transferMode = _configService.CopyUsingHardlinks ? TransferMode.HardLinkOrCopy : TransferMode.Copy;
+
+            // TODO: check for readonly / downloaditem etc, adjust transferMode
+            if (extraFilePath != newFileName)
```
I would like to avoid this check, but if source and destination are the same, `TransferFile()` will throw an exception.

tty418 replied:
Changed this comparison to work on two DirectoryInfo instances, not perfect but it covers more cases :)

tty418 commented on src/NzbDrone.Core/Datastore/Migration/075_relax_not_null_constraints_extra_files.cs:
```diff
@@ -0,0 +1,17 @@
+
+namespace NzbDrone.Core.Datastore.Migration
+{
+    [Migration(075)]
```
Done, also rebased on develop.

tty418 commented:
@mynameisbogdan (or anyone reviewing): as a temporary measure I'm considering adding an option to configure the list of supported extensions for the album extras. The mid-term plan is, as before, to add configuration through the UI in a follow-up PR. But until then I could enable users or people willing to test the feature specify which extensions should be handled.

The easiest thing to do would be to read an environment variable, and then override the hardcoded list:
`if (Environment.GetEnvironmentVariable("LIDARR_EXPERIMENTAL_EXTRA_FILE_EXTENSIONS") != null) { ... }`
Or use the configuration provider and inject `IOptions<LidarrExperimentalFoo>` - cleaner, but more involved and probably unnecessary for a temporary setting.

This would resolve the issue mentioned here: https://github.com/Lidarr/Lidarr/issues/484#issuecomment-1915957008

Qstick commented on src/NzbDrone.Core/Datastore/Migration/076_relax_not_null_constraints_extra_files.cs:
```diff
@@ -0,0 +1,17 @@
+    {
+        protected override void MainDbUpgrade()
+        {
+            Rename.Column("TrackFileId").OnTable("ExtraFiles").To("TrackFileIdLegacy");
+            Alter.Table("ExtraFiles").AddColumn("TrackFileId").AsInt64().Nullable();
+            Execute.Sql("UPDATE \"ExtraFiles\" SET \"TrackFileId\" = \"TrackFileIdLegacy\"");
+            Delete.Column("TrackFileIdLegacy").FromTable("ExtraFiles");
```
Does AlterColumn not work correctly in this case?

```
Alter.Table("ExtraFiles").AlterColumn("TrackFileId").AsInt32().Nullable();
```

tty418 replied:
Since SQLite doesn't support type changes I wrongly assumed that this needs to be manually handled in the migration. Apparently Fluent Migrator managed the change correctly. Fixed 👍

Qstick commented on src/NzbDrone.Core/Extras/ExtraService.cs:
```diff
@@ -48,9 +52,32 @@ public class ExtraService : IExtraService,
+            var filtered = albumFiles.Where(x => !importedTracks.Select(t => t.Item.Path).Contains(x));
+
+            // TODO: grab extensions from config?
+            var extensions = new List<string> { ".cue", ".jpg", ".png", ".nfo", ".log", ".txt" };
```
_configService.ExtraFileExtensions

or a new setting if we think it's needed (existing for Artist and new for Album extensions)

tty418 replied:
I have two questions regarding this suggestion:
1. The default value currently is `srt`, which is probably coming from the TV Arrs. 
-> Shall we define a list of suitable extensions, e.g. the ones I did my tests with and hardcoded above?
2. The setting is currently off by default.
-> I was hoping that this feature can be instead on / opt out.

Especially wrt the second point - the impact of downloading dozens of albums and finding out afterwards that none of the usual extra files (nfos, covers, cue's) are there is huge. I literally had to stop my noise file acquisition for weeks until I figured out what's happening and came up with this PR. I'm still slowly going through my spreadsheet and rebuilding artist by artist.

With that said, I assumed that this should be a different setting, which does not seem to make sense. Putting this under the umbrella of "Extra Files" is fine.

Qstick replied:
1. Yes, let's change the default to something that makes sense. SRT doesn't (I'm actually surprised it's gone this long)
2. I'm not opposed to on by default if we think that makes sense for music.

tty418 replied:
Alright, I made the changes. The defaults are as following: 
- Import extra files on by default;
- Extra file extensions: `log, cue, nfo, jpg, jpeg, png`

Regarding sensible defaults for the extensions:
- Must-haves are `.log` and `.cue`: highly valued in at least one particular music _community_ :)
- Nice to haves: `.nfo` (release info)
- Maybe: `.pdf` and `.txt` (release-specific, some include such extra files or docs)
- Whatever: cover files (so your jpg's, png's etc), this is recoverable information

I could add all of these, but I didn't want to make the list too long. Let me know what you think.

Qstick commented on src/NzbDrone.Core/Extras/ExtraService.cs:
```diff
@@ -48,9 +52,32 @@ public class ExtraService : IExtraService,
+        /// Import extra files belonging to an album. Track-specific extras will be ignored.
+        /// </summary>
+        /// <param name="importedTracks">The imported tracks in the source dir.</param>
+        /// <param name="destinationDir">Destination dir for the album extras</param>
```
Blitz the method comments, we try and avoid it with the codebase (our style)

tty418 replied:
Done.

Qstick commented on src/NzbDrone.Core/Extras/Files/AlbumExtraFileManager.cs:
```diff
@@ -0,0 +1,166 @@
+
+        public IEnumerable<ExtraFile> ImportAlbumExtras(Artist artist, int albumId, IEnumerable<string> extraFilePaths, string targetDirectory)
+        {
+            var cfg = _namingConfigService.GetConfig();
```
`cfg` to `namingConfig`

tty418 replied:
Done.

tty418 commented on src/NzbDrone.Core/Datastore/Migration/076_relax_not_null_constraints_extra_files.cs:
```diff
@@ -0,0 +1,17 @@
+
+namespace NzbDrone.Core.Datastore.Migration
+{
+    [Migration(076)]
```
Updated attribute and migration file name.

Qstick commented on src/NzbDrone.Core/Datastore/Migration/078_relax_not_null_constraints_extra_files.cs:
```diff
@@ -0,0 +1,14 @@
+    {
+        protected override void MainDbUpgrade()
+        {
+            Alter.Table("ExtraFiles").AlterColumn("TrackFileId").AsInt64().Nullable();
```
Is the change from int32 to int64 intentional?

tty418 replied:
Not really, I did it on autopilot. Reverted to Int32
Underlying column types are, according to the SQLite documentation, int64 already.

tty418 commented:
It turned out that extra file imports were handled only in the simplest case. Multi-CD releases and basically anything different than a single directory for all files resulted in files not being found, or being moved to the incorrect target directory on import. 

All problems should be solved now. At the price of quite a bit of path wrangling and extra route matching code. I tried to keep things simple and readable, with the occasional comment here and there to explain why something is done in a particular way.

The tests, both setup and the way I organized them, also needed lots of adjustments. I am happy with the end result, but faking and mocking a file system with recursive file search hurts. But without the tests I would've never made it.

The unit tests cover all **supported directory structure** variants:
- tracks at root, extras at any dir under root (search recursively)
- tracks in a single subdir, extras only under subdir (otherwise its unsafe)
- tracks in multiple subdirs (CD1, CD2 etc), extras at root or any subdir (recursively)

@mynameisbogdan  @Qstick 
TL DR:
- large parts of the PR were rewritten
- importing downloads and manually / in place works:
  - High confidence that it works well on Windows / while debugging with VS
  - Cross-platform tests caught a few issues early on
  - As soon as a Docker Image is ready I'll quickly test on linux

tty418 commented:
> FYI just did a rebase in #4526

I rebased develop and fixed the issue reported by Sonar. I also used the opportunity to extract some of the file matching code: 
https://github.com/Lidarr/Lidarr/pull/4370/files/59b984f621f72a5516c541a61bb76740ebdacee2..9ae4b1fb16e7e81aa57c18d4fa2b1dd44f1d6bbe

Let me know if the code is hard to follow somewhere, or if you find issues in the PR :)

tty418 commented:
> can you squash 36 commits into a single one as well?

Squashed 👍

rodrigorodrigo commented:
just coming here to say I've been using this PR for a while (ghcr.io/linuxserver-labs/prarr:lidarr-extras-for-albums) and it's been working amazingly well. +1 on reviewing and adding this feature, finally, on lidarr.

augustuen commented on src/NzbDrone.Test.Common/TestBase.cs:
```diff
@@ -103,7 +103,7 @@ protected string TempFolder
         public void TestBaseSetup()
         {
-            GetType().IsPublic.Should().BeTrue("All Test fixtures should be public to work in mono.");
+            GetType().Should().Match(t => t.IsPublic || t.IsNestedPublic, "All Test fixtures should be public to work in mono.");
```
Why is this included in this PR?
